MAINT: Add inline type hintings for stats.qmc

V0lantis opened the pull request:
Following @tupui and @rgommers [comment](https://github.com/scipy/scipy/pull/13576#issuecomment-813577228) on #13576, I am submitting this PR to remove `_qmc.pyi` to add inline type hinting inside `_qmc.py`
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
https://docs.scipy.org/doc/numpy/dev/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
None

#### What does this implement/fix?
<!--Please explain your changes.-->
This implementation remove a type files (`qmc.pyi`) and replace it by inline typing directly inside `_qmc.py`

V0lantis commented:
I wonder what is the problem : 
```bash
$ python -u runtests.py --mypy
Building, see build.log...
Build OK (0:00:52.238457 elapsed)
Success: no issues found in 679 source files
$ python --version            
Python 3.9.2
```

tupui commented:
Thanks! I will pull the branch and check tmr.

Do you plan to also do the same for the discrepancy? I thought we were talking about this instead in the other PR. But this is also wanted 👍

V0lantis commented:
Oh yes, of course ! I forgot that it was also possible in cython files.

V0lantis commented on scipy/stats/_qmc.py:
```diff
@@ -389,7 +416,7 @@ def n_primes(n):
 
     """
-    primes = [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59,
+    primes = np.array([2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59,
```
I had to add `np.array` at the front because n_primes is returning a `np.ndarray`

V0lantis commented on scipy/stats/_qmc.py:
```diff
@@ -265,12 +288,13 @@ def discrepancy(sample, iterative=False, method="CD", workers=1):
 
     workers = int(workers)
     if workers == -1:
-        workers = os.cpu_count()
-        if workers is None:
+        cpu_count = os.cpu_count()
```
I created here this variable since `os.cpu_count` is returning `Optional[int]` and `workers` is a plain `int`. Since a couple of lines before, I use `workers = int(workers)`, and int is receiving `Union[str, bytes, SupportsInt, _SupportsIndex, _SupportsTrunc]` and not `None`, I couldn't change workers type into `Optional[int]`

tupui replied:
It was not working to have `workers: Optional[int] = int(workers)`?

V0lantis commented on scipy/stats/_qmc.py:
```diff
@@ -1244,7 +1297,7 @@ def _standard_normal_samples(self, n=1):
             # apply inverse transform
             # (values to close to 0/1 result in inf values)
-            return stats.norm.ppf(0.5 + (1 - 1e-10) * (samples - 0.5))
+            return stats.norm.ppf(0.5 + (1 - 1e-10) * (samples - 0.5)) # type: ignore[attr-defined]
```
Had to add this because of python/mypy#7182. Otherwise, I had this error : 
`scipy/stats/_qmc.py:1300: error: Module has no attribute "norm"  [attr-defined]`

V0lantis commented on scipy/stats/_qmc.py:
```diff
@@ -1183,13 +1236,13 @@ def __init__(self, mean, cov=None, cov_root=None, inv_transform=True,
             engine_dim = d
         if engine is None:
-            self.engine = Sobol(d=engine_dim, scramble=True, seed=seed)
+            self.engine = Sobol(d=engine_dim, scramble=True, seed=seed) # type: QMCEngine
```
Had to add this to tel Mypy that `self.engine` has `QMCEngine` type. See scipy/stats/_qmc.py:1300: error: Module has no attribute "norm"  [attr-defined]

V0lantis reviewed (commented):
I had some comment on line that may not be clear from first sight. I hope this time that all tests will pass

tupui commented:
Seems like we cannot import `numpy.typing`

```python

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/vsts/work/1/s/build/testenv/lib/python3.7/site-packages/scipy/signal/__init__.py", line 310, in <module>
    from ._peak_finding import *
  File "/home/vsts/work/1/s/build/testenv/lib/python3.7/site-packages/scipy/signal/_peak_finding.py", line 8, in <module>
    from scipy.stats import scoreatpercentile
  File "/home/vsts/work/1/s/build/testenv/lib/python3.7/site-packages/scipy/stats/__init__.py", line 443, in <module>
    from . import qmc
  File "/home/vsts/work/1/s/build/testenv/lib/python3.7/site-packages/scipy/stats/qmc.py", line 234, in <module>
    from ._qmc import *
  File "/home/vsts/work/1/s/build/testenv/lib/python3.7/site-packages/scipy/stats/_qmc.py", line 30, in <module>
    import numpy.typing as npt
ModuleNotFoundError: No module named 'numpy.typing'

```

tupui commented on scipy/stats/_qmc.pyi:
```diff
@@ -1,138 +0,0 @@
-
-import numpy as np
-
-if TYPE_CHECKING:
-    import numpy.typing as npt
```
This is needed for NumPy < 1.20

V0lantis replied:
It is not used in [`SciPy/stats/_entropy.py`](https://github.com/scipy/scipy/blob/master/scipy/stats/_entropy.py) where there are also type annotations

tupui replied:
Sure because it uses directly in the type hints `np.typing` then this is not evaluated at runtime. But here you import `numpy.typing` so this will only work for NumPy > 1.2. Hence the failing jobs.

V0lantis replied:
Ok will try this solution then

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -64,7 +78,11 @@ def check_random_state(seed=None):
 
 
-def scale(sample, l_bounds, u_bounds, reverse=False):
+def scale(
+        sample: np.typing.ArrayLike,
+        l_bounds: np.typing.ArrayLike,
+        u_bounds: np.typing.ArrayLike,
+        reverse: bool = False) -> np.ndarray:
```
Careful about the extra spaces and the ending comma. It should be like this
```suggestion
def scale(
    sample: np.typing.ArrayLike,
    l_bounds: np.typing.ArrayLike,
    u_bounds: np.typing.ArrayLike,
    reverse: bool = False
) -> np.ndarray:
```

ilayn replied:
Can we at least do `from numpy.typing import ArrayLike`. That's already lots of repetition.

tupui replied:
True. He removed it because of the import error. But I guess it was a better choice in the first place to use the following (as it was in the pyi)
```python
if TYPE_CHECKING:
    import numpy.typing as npt
```

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -964,7 +1006,11 @@ class Sobol(QMCEngine):
+    def __init__(
+            self, d: int, scramble: bool = True,
+            seed: SeedType = None
+
```
```suggestion
```

ilayn commented on scipy/stats/_qmc.py:
```diff
@@ -1,33 +1,47 @@
            'QMCEngine', 'Sobol', 'Halton', 'LatinHypercube',
            'MultinomialQMC', 'MultivariateNormalQMC']
 
+_IntegerType = Union[int, np.integer]
```
Call these `IntNumber` and `DecimalNumber` or something more non-code? It gets really crowded towards the end of this file.

V0lantis replied:
Changes are in 10ba2affd279d2bbce932cdeadfbbe6340876eea. Thank you @ilayn

ilayn commented on scipy/stats/_qmc.py:
```diff
@@ -1141,8 +1187,13 @@ class MultivariateNormalQMC(QMCEngine):
-    def __init__(self, mean, cov=None, cov_root=None, inv_transform=True,
-                 engine=None, seed=None):
+    def __init__(
+            self, mean: np.typing.ArrayLike, cov: Optional[np.typing.ArrayLike] = None,
```
For example here. I have no idea what is happening here and I don't think any IDE can help me out here to understand what the signature is. 

If I remember correctly for keywords it will automatically register as Optional.

tupui replied:
This is ok for an IDE. We need the `Optional` for mypy as it's not enough that the default is `None`. If that was your concern.

ilayn replied:
Shouldn't this be eligible for https://mypy.readthedocs.io/en/stable/runtime_troubles.html#using-x-y-syntax-for-unions ?

ilayn replied:
So we can write `ArrayLike | None` maybe instead of this garbled stuff?

tupui replied:
Union? I think there might be a misunderstanding here. There are 2 parameters here `mean` and `cov`

tupui replied:
`Optional[npt.ArrayLike]` or without the `npt` but not sure about the `|`. Is this supposed to replace the Optional too?

tupui replied:
Apparently yes (https://www.python.org/dev/peps/pep-0604/#proposal). Not sure about the best practices we want here. But as we will still need to write the default for python itself, IMO this will look redundant as `ArrayLike | None = None`

V0lantis replied:
Didn't change it then

V0lantis commented on scipy/stats/_qmc.py:
```diff
@@ -263,7 +285,7 @@ def discrepancy(sample, iterative=False, method="CD", workers=1):
         raise ValueError("Sample is not in unit hypercube")
 
-    workers = int(workers)
+    workers: Optional[int] = int(workers)
```
Ok so this is not possible : `scipy/stats/_qmc.py:288: error: Name 'workers' already defined on line 168  [no-redef]`

tirthasheshpatel reviewed (changes requested):
Made some suggestions on failing CI.

V0lantis commented on scipy/stats/_qmc.py:
```diff
@@ -800,11 +837,15 @@ class LatinHypercube(QMCEngine):
         self.centered = centered
 
-    def random(self, n=1):
+    def random(self, n: int = 1) -> Union[np.number, np.ndarray]:
```
The issue is if someone uses this function and expects a float, he will have a type error in his IDE (because the IDE considers that it should be an `np.ndarray`). Am I correct?

tirthasheshpatel replied:
But why would someone expect a float? From my understanding, the output is always a `ndarray`. So, it will correctly show a type error, right?

tupui replied:
Agreed with @tirthasheshpatel, all random are returning `np.ndarray`. The original types from `_qmc.pyi` should be correct and not require much adjustment as we already iterated over this.

V0lantis replied:
I return to the previous state in 7a599c601292307a8f438b810b837af976369122 and add a ignore type at the return value

V0lantis commented on scipy/stats/_qmc.py:
```diff
@@ -263,7 +285,7 @@ def discrepancy(sample, iterative=False, method="CD", workers=1):
         raise ValueError("Sample is not in unit hypercube")
 
-    workers = int(workers)
+    workers: Optional[int] = int(workers)
```
I changed it in d043fce28a901fb9be323cc1048ea4ebc6dd99f6. Is that ok for you?

tirthasheshpatel replied:
Looks good!

V0lantis reviewed (commented):
Thank you the 3 of you for your comment. I reviewed them in the past commits

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -1,33 +1,52 @@
            'QMCEngine', 'Sobol', 'Halton', 'LatinHypercube',
            'MultinomialQMC', 'MultivariateNormalQMC']
 
+IntNumber = Union[int, np.integer]
+DecimalNumber = Union[float, np.floating]
```
These should be used as before through the file. Here is some background https://github.com/scipy/scipy/pull/10844#pullrequestreview-559980027

V0lantis replied:
The issue is, with the old types (in `_qmc.pyi`) we have got a lot of types errors when putting them into inline types : 
```Shell
scipy/stats/_qmc.py:400: error: Argument 1 to "ones" has incompatible type "Union[int, number[Any]]"; expected "Union[int, Sequence[int]]"  [arg-type]
scipy/stats/_qmc.py:435: error: Slice index must be an integer or None  [misc]
scipy/stats/_qmc.py:440: error: Slice index must be an integer or None  [misc]
scipy/stats/_qmc.py:492: error: Argument 1 to "zeros" has incompatible type "Union[int, integer[Any]]"; expected "Union[int, Sequence[int]]"  [arg-type]
scipy/stats/_qmc.py:773: error: Incompatible types in assignment (expression has type "Union[int, number[Any]]", variable has type "int")  [assignment]
scipy/stats/_qmc.py:774: error: Argument 1 to "reshape" of "_ArrayOrScalarCommon" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
scipy/stats/_qmc.py:774: error: Argument 2 to "reshape" of "_ArrayOrScalarCommon" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
scipy/stats/_qmc.py:873: error: Argument 1 to "range" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
scipy/stats/_qmc.py:878: error: Incompatible types in assignment (expression has type "Union[int, number[Any]]", variable has type "int")  [assignment]
scipy/stats/_qmc.py:1028: error: Argument 1 to "zeros" has incompatible type "Tuple[Union[int, integer[Any]], Any]"; expected "Union[int, Sequence[int]]"  [arg-type]
scipy/stats/_qmc.py:1032: error: Argument 1 to "zeros" has incompatible type "Union[int, integer[Any]]"; expected "Union[int, Sequence[int]]"  [arg-type]
scipy/stats/_qmc.py:1071: error: Argument 1 to "empty" has incompatible type "Tuple[Union[int, integer[Any]], Union[int, integer[Any]]]"; expected "Union[int, Sequence[int]]"  [arg-type]
scipy/stats/_qmc.py:1075: error: Unsupported operand types for & ("int" and "number[Any]")  [operator]
scipy/stats/_qmc.py:1075: error: Unsupported operand types for & ("integer[Any]" and "number[Any]")  [operator]
scipy/stats/_qmc.py:1075: note: Both left and right operands are unions
scipy/stats/_qmc.py:1084: error: Slice index must be an integer or None  [misc]
scipy/stats/_qmc.py:1089: error: Incompatible types in assignment (expression has type "Union[int, number[Any]]", variable has type "int")  [assignment]
scipy/stats/_qmc.py:1112: error: Unsupported left operand type for & ("number[Any]")  [operator]
scipy/stats/_qmc.py:1112: note: Both left and right operands are unions
scipy/stats/_qmc.py:1120: error: Argument 1 to "random" of "Sobol" has incompatible type "Union[Any, number[Any]]"; expected "Union[int, integer[Any]]"  [arg-type]
scipy/stats/_qmc.py:1303: error: unused 'type: ignore' comment
scipy/stats/_qmc.py:1304: error: Module has no attribute "norm"  [attr-defined]
scipy/stats/_qmc.py:1313: error: Argument 1 to "reshape" of "_ArrayOrScalarCommon" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
scipy/stats/_qmc.py:1315: error: Slice index must be an integer or None  [misc]
```

tupui replied:
@rgommers apparently mypy complains about `_IntegerType = Union[int, np.integer]` and the same for float. Is it fine then to use plain int?

rgommers replied:
That's odd, it shouldn't matter if the annotation is inline or in a stub file. Just to make sure: this is with the most recent NumPy version?

```

scipy/stats/_qmc.py:492: error: Argument 1 to "zeros" has incompatible type "Union[int, integer[Any]]"; expected "Union[int, Sequence[int]]"  [arg-type]
```

The current definition of `_ShapeLike` is `Union[int, Sequence[int]]`, so that indeed doesn't accept numpy integer types. That may be an oversight - @BvB93 can you comment?

BvB93 replied:
That definitely seems to be an oversight, yes. At a quick glance `int` either has to be replaced with the `typing.SupportsIndex` protocol or, alternatively, a union of `int` and `np.integer`.

BvB93 replied:
FYI, a PR addressing the `_ShapeLike` issue is currently up at https://github.com/numpy/numpy/pull/18778.

rgommers replied:
Thanks Bas!

Okay, so we want to keep `_IntegerType` the same here. It works with numpy ~master~ main now. It's just a little annoying that CI will fail. Or at least it should - it's not failing now, and we should figure out why cause apparently it's broken. 

We can then switch `.github/workflows/linux.yml` to use numpy nightlies to test, because this won't be the last time we need a fix in numpy.

tupui replied:
Do we want to use this `_IntegerType ` for all future type hints or `int` in SciPy? If yes we should move this to `_utils` or something else no? I am still confused about the need of this special type over `int` TBH.

Are you saying we should switch to NumPy nightlies here?

rgommers replied:
> Do we want to use this `_IntegerType` for all future type hints or int in SciPy? If yes we should move this to `_utils` or something else no? 

We should move it to `numpy.typing` at some point.

> I am still confused about the need of this special type over int TBH.

It's because of the unfortunate existence of numpy array scalars - `np.int64(1)` is not a Python integer, but acts like one.

> Are you saying we should switch to NumPy nightlies here?

Better in a separate PR, unless it's necessary here because CI is failing.

tupui replied:
I am asking because we have also types in `differential_entropy` and we have used plain `int` there. Maybe we should communicate about new guidelines on how to type hint before we start adding "wrong" things.

V0lantis replied:
Unfortunatly, the nightly release is from 4 days ago (https://anaconda.org/scipy-wheels-nightly/numpy/files?sort=time&sort_order=desc). I am trying to build with @BvB93 modifications right now (so the main release).

rgommers replied:
> Maybe we should communicate about new guidelines on how to type hint before we start adding "wrong" things.

We're going to find things like this as we go. How about we keep a wiki page with pointers for now? I feel like we're not at the point where it's easy/productive to write a section of the contributing guide yet.

V0lantis replied:
Just to say that with the news modifications, we have new `MyPy` errors unrelated to this PR:
```bash

scipy/special/orthogonal.pyi:273: error: Variable "numpy.poly1d" is not valid as a type  [valid-type]
scipy/special/orthogonal.pyi:273: note: See https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases
scipy/special/orthogonal.pyi:273: error: Invalid base class "numpy.poly1d"  [misc]
scipy/fft/_pocketfft/tests/test_real_transforms.py:256: error: Invalid index type "Tuple[partial[Any], Type[int], int]" for "Dict[Tuple[partial[Any], Type[floating[Any]], int], int]"; expected type "Tuple[partial[Any], Type[floating[Any]], int]"  [index]
```

I must admit I feel a little bit lost. What should be the next steps ?

tupui replied:
Agreed @rgommers. I am not sure I can help here as I am struggling to get all these sorry. I know about normal typing but not NumPy specifics like these. I can review the document though if it can help.

BvB93 replied:
> Just to say that with the news modifications, we have new MyPy errors unrelated to this PR:
>
> ```
> scipy/special/orthogonal.pyi:273: error: Variable "numpy.poly1d" is not valid as a type  [valid-type]
> scipy/special/orthogonal.pyi:273: note: See https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases
> scipy/special/orthogonal.pyi:273: error: Invalid base class "numpy.poly1d"  [misc]
> ```

Great, it appears that mypy does not allow the subclassing of objects annotated as `Type[Any]`, 
only explicitly defined classes. This is annoying, but it's something that we can fix in numpy with
a small followup on https://github.com/numpy/numpy/pull/18765/commits/b9d9e03014aaa4d7b6f06a89ceae06317a0012a3.

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -147,7 +171,11 @@ def scale(sample, l_bounds, u_bounds, reverse=False):
 
 
-def discrepancy(sample, iterative=False, method="CD", workers=1):
+def discrepancy(
+        sample: npt.ArrayLike,
+        iterative: bool = False,
+        method: str = "CD",
+        workers: int = 1) -> DecimalNumber:
```
```suggestion
def discrepancy(
    sample: npt.ArrayLike,
    iterative: bool = False,
    method: str = "CD",
    workers: int = 1
) -> DecimalNumber:
```

BvB93 replied:
Same comment as above: try to avoid returning unions.
Based on the annotations in https://github.com/scipy/scipy/blob/master/scipy/stats/_qmc_cy.pyi it seems to return a normal `float`.

```suggestion
        workers: int = 1) -> float:
```

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -289,7 +317,10 @@ def discrepancy(sample, iterative=False, method="CD", workers=1):
 
-def update_discrepancy(x_new, sample, initial_disc):
+def update_discrepancy(
+        x_new: npt.ArrayLike,
+        sample: npt.ArrayLike,
+        initial_disc: float) -> DecimalNumber:
```
```suggestion
    x_new: npt.ArrayLike,
    sample: npt.ArrayLike,
    initial_disc: float
) -> DecimalNumber:
```

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -414,7 +445,12 @@ def n_primes(n):
 
 
-def van_der_corput(n, base=2, start_index=0, scramble=False, seed=None):
+def van_der_corput(
+        n: int,
+        base: IntNumber = 2,
+        start_index: IntNumber = 0,
+        scramble: bool = False,
+        seed: SeedType = None) -> np.ndarray:
```
```suggestion
def van_der_corput(
    n: int,
    base: IntNumber = 2,
    start_index: IntNumber = 0,
    scramble: bool = False,
    seed: SeedType = None
) -> np.ndarray:
```

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -563,14 +599,18 @@ class QMCEngine(ABC):
 
     @abstractmethod
-    def __init__(self, d, seed=None):
+    def __init__(
+            self,
+            d: int,
+            seed: SeedType = None
+    ) -> None:
```
```suggestion
    def __init__(
        self,
        d: int,
        seed: SeedType = None
    ) -> None:
```

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -700,13 +740,16 @@ class Halton(QMCEngine):
     """
 
-    def __init__(self, d, scramble=True, seed=None):
+    def __init__(
+            self, d: int, scramble: bool = True,
+            seed: SeedType = None
+    ) -> None:
```
```suggestion
    def __init__(
        self, d: int, scramble: bool = True,
        seed: SeedType = None
    ) -> None:
```

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -800,11 +843,15 @@ class LatinHypercube(QMCEngine):
     """
 
-    def __init__(self, d, centered=False, seed=None):
+    def __init__(
+            self, d: int, centered: bool = False,
+            seed: SeedType = None
+
+    ) -> None:
```
```suggestion
    def __init__(
        self, d: int, centered: bool = False,
        seed: SeedType = None
    ) -> None:
```

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -964,7 +1012,10 @@ class Sobol(QMCEngine):
     MAXBIT = _MAXBIT
 
-    def __init__(self, d, scramble=True, seed=None):
+    def __init__(
+            self, d: int, scramble: bool = True,
+            seed: SeedType = None
+    ) -> None:
```
```suggestion
    def __init__(
        self, d: int, scramble: bool = True,
        seed: SeedType = None
    ) -> None:
```

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -1141,8 +1192,13 @@ class MultivariateNormalQMC(QMCEngine):
 
-    def __init__(self, mean, cov=None, cov_root=None, inv_transform=True,
-                 engine=None, seed=None):
+    def __init__(
+            self, mean: npt.ArrayLike, cov: Optional[npt.ArrayLike] = None,
+            cov_root: Optional[npt.ArrayLike] = None,
+            inv_transform: bool = True,
+            engine: Optional[QMCEngine] = None,
+            seed: SeedType = None
+    ) -> None:
```
```suggestion
    def __init__(
        self, mean: npt.ArrayLike, cov: Optional[npt.ArrayLike] = None,
        cov_root: Optional[npt.ArrayLike] = None,
        inv_transform: bool = True,
        engine: Optional[QMCEngine] = None,
        seed: SeedType = None
    ) -> None:
```

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -1283,7 +1339,10 @@ class MultinomialQMC(QMCEngine):
     """
 
-    def __init__(self, pvals, engine=None, seed=None):
+    def __init__(
+            self, pvals: npt.ArrayLike, engine: Optional[QMCEngine] = None,
+            seed: SeedType = None
+    ) -> None:
```
```suggestion
    def __init__(
        self, pvals: npt.ArrayLike, engine: Optional[QMCEngine] = None,
        seed: SeedType = None
    ) -> None:
```

tupui reviewed (changes requested):
I think you might have an autoformatting setting to adjust to avoid this double indentation issue.

V0lantis commented:
> I think you might have an autoformatting setting to adjust to avoid this double indentation issue.

Yes I agree, but shouldn't it be like this ? (https://www.python.org/dev/peps/pep-0008/#indentation) :

```python
# Add 4 spaces (an extra level of indentation) to distinguish arguments from the rest.
def long_function_name(
        var_one, var_two, var_three,
        var_four):
    print(var_one)
```

tupui commented:
> > I think you might have an autoformatting setting to adjust to avoid this double indentation issue.
> 
> Yes I agree, but shouldn't it be like this ? (https://www.python.org/dev/peps/pep-0008/#indentation) :
> 
> ```python
> # Add 4 spaces (an extra level of indentation) to distinguish arguments from the rest.
> def long_function_name(
>         var_one, var_two, var_three,
>         var_four):
>     print(var_one)
> ```

I am puzzled here! Thanks for pointing this out. I will have to check. I really thought things like this were not correct

```python
# Hanging indents should add a level.
foo = long_function_name(
    var_one, var_two,
    var_three, var_four)
```
But should be 
```python
# Hanging indents should add a level.
foo = long_function_name(
    var_one, var_two,
    var_three, var_four
)
```

tupui commented:
According to flake8 (E125), both this and my suggestion are fine. As you wish then if both are evaluated as valid.

V0lantis commented:
> According to flake8 (E125), both this and my suggestion are fine. As you wish then if both are evaluated as valid. (Feel free to resolve)

Thank you @tupui. Unfortunately I wasn't able to put the same types as before because they raise a lot of errors which were't there with `_qmc.pyi` file.

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -1,33 +1,52 @@
 
 # Based on scipy._lib._util.check_random_state
-def check_random_state(seed=None):
+def check_random_state(seed: SeedType = None) -> Union[np.random.Generator,
```
This looks scary... But ok we can do this here as in general we don't have much unions.

V0lantis commented on scipy/stats/_qmc.py:
```diff
@@ -147,7 +171,11 @@ def scale(sample, l_bounds, u_bounds, reverse=False):
+def discrepancy(
+        sample: npt.ArrayLike,
+        iterative: bool = False,
+        method: str = "CD",
```
Since there is a disagreement here, I am waiting for more input :-)

BvB93 replied:
Unless the type of this `method` parameter is very common in scipy I'd suggest to just directly assign it to the `method` parameter without the inclusion of any type aliases.

tupui replied:
To clarify what I ment. In a lot of stats functions we have a literal (`alternative`). I am suggesting we could use the same logic in such cases, hence have a common helper to import `literal` and other things. Otherwise we will add quite some boilerplate to a lot of files.

V0lantis replied:
Updated in f9eb736c78310d2f294d5b2b558d849a151a99e7. I didn't put the Literal in `_util` though, since the method arguments are quite specific to the discrepancy function

V0lantis commented:
I have updated Nightly CPython workflow in `.github/workflow/linux.yml`'s  with python version 3.9. It is just for testing purposes and I intend to remove my changes where all the tests are passing. 
Basically, I just do a pull on the main numpy's branch. It is just a test, I am not a well experienced with those kinds of things. The goal is to see the latest changes on NumPy types. 

> We can then switch .github/workflows/linux.yml to use numpy nightlies to test, because this won't be the last time we need a fix in numpy.

@rgommers I would be very interested in "getting my hands dirty" and try to make a PR upon Scipy's workflows to integrate NumPy's Nightly release. But I just saw that the Nightly Released was due every Sunday. Wouldn't it be better to do it rather every night ? For example, @BvB93 latest changes haven't been released yet in the nightly released. If we have to wait for every Sunday to test if our changes are passing, it is rather long. 
Or is my method more relevant, i.e. pulling NumPy's main branch and building it directly in the workflow ?

rgommers commented:
> @rgommers I would be very interested in "getting my hands dirty" and try to make a PR upon Scipy's workflows to integrate NumPy's Nightly release.

That'd be great, thanks.

> But I just saw that the Nightly Released was due every Sunday. Wouldn't it be better to do it rather every night ? 

The reason is that we have limited CI over on https://github.com/MacPython/numpy-wheels after TravisCI became less generous. 

> For example, @BvB93 latest changes haven't been released yet in the nightly released. If we have to wait for every Sunday to test if our changes are passing, it is rather long.
> Or is my method more relevant, i.e. pulling NumPy's main branch and building it directly in the workflow ?

If we know we need a change and don't want to wait for the new nightly (which shouldn't happen _that_ often), we can always trigger a one-off build on https://github.com/MacPython/numpy-wheels. That'd be better than building numpy from source - that's a bit slow, if there's something wrong with it it'll then break both numpy and scipy CI.

V0lantis commented:
What should we do for the errors which are not related to my the PR, following the changes from @BvB93. Should I first submit a PR which is going to resolve them ?
Erros : 
<details>

```bash
scipy/special/orthogonal.pyi:273: error: Variable "numpy.poly1d" is not valid as a type  [valid-type]
22
scipy/special/orthogonal.pyi:273: note: See https://mypy.readthedocs.io/en/latest/common_issues.html#variables-vs-type-aliases
23
scipy/special/orthogonal.pyi:273: error: Invalid base class "numpy.poly1d"  [misc]
24
scipy/fft/_pocketfft/tests/test_real_transforms.py:256: error: Invalid index type "Tuple[partial[Any], Type[int], int]" for "Dict[Tuple[partial[Any], Type[floating[Any]], int], int]"; expected type "Tuple[partial[Any], Type[floating[Any]], int]"  [index]

</details>

Also, there is an [error](https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=11527&view=logs&j=ef553e3c-aa16-5ea0-cd50-e42a8acf453d&t=271890b8-652f-5dc0-a332-d417b8e08791&l=37030) in Azure tests : `AttributeError: module 'numpy.random' has no attribute 'Generator'`. Is this related to some workflows failed ?

rgommers commented:
> What should we do for the errors which are not related to my the PR, following the changes from @BvB93. Should I first submit a PR which is going to resolve them ?

A separate PR to fix those sounds great.

> Also, there is an [error](https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=11527&view=logs&j=ef553e3c-aa16-5ea0-cd50-e42a8acf453d&t=271890b8-652f-5dc0-a332-d417b8e08791&l=37030) in Azure tests : `AttributeError: module 'numpy.random' has no attribute 'Generator'`. Is this related to some workflows failures ?

This will be due to an old numpy version. You cannot use `Generator` unconditionally, it was introduced in numpy 1.17 and we still have 1.16.5 as minimum.

BvB93 commented on scipy/stats/_qmc.py:
```diff
@@ -1,29 +1,61 @@
+IntNumber = Union[int, np.integer]
+DecimalNumber = Union[float, np.floating]
+
+SeedType = Optional[Union[IntNumber, np.random.Generator,
+                          np.random.RandomState]]
```
Moving these lines to the `if TYPE_CHECKING:` block should get rid of the numpy 1.16-related `AttributeError` (https://github.com/scipy/scipy/pull/13833#issuecomment-821263620).

V0lantis replied:
Should I move all types declaration in the `if` statement, including `IntNumber` and every other types I wish to declare ?

BvB93 replied:
For statements that can be executed just fine during runtime this won't be (strictly) necessary. 
It can still be done however, but this would be more of a stylistic choice; one I don't have any strong opinions about.

V0lantis commented on scipy/stats/_qmc.py:
```diff
@@ -401,7 +446,7 @@ def n_primes(n):
               769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857,
               859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947,
-              953, 967, 971, 977, 983, 991, 997][:n]
+              953, 967, 971, 977, 983, 991, 997])[:n]
```
@BvB93 Do you know how to to type `n` ? 
For now, we have the following error :
```bash
scipy/stats/_qmc.py:449: error: Slice index must be an integer or None  [misc]
scipy/stats/_qmc.py:454: error: Slice index must be an integer or None  [misc]
```
because `builtin.slice` is not accepting other types than `int` or `None` (see python/typing#159)
Apparently (https://github.com/python/mypy/issues/2410#issuecomment-260159393), NumPy `slice` method accepts other types. Do you know how to correct this ?

BvB93 replied:
Given the current mypy issue, I'm afraid the only way to deal with this false positive is to add a `# type: ignore[misc]` comment.

V0lantis commented:
Some Updates 
New errors are appeared out of nowhere. I wonder if it is due to new commit on NumPy's main branch...
```bash
scipy/stats/_constants.py:9: error: Too many arguments for "finfo"  [call-arg]
21
scipy/stats/_constants.py:12: error: Too many arguments for "finfo"  [call-arg]
22
scipy/stats/_constants.py:19: error: Too many arguments for "finfo"  [call-arg]
23
scipy/special/orthogonal.pyi:287: error: Self argument missing for a non-static method (or an invalid type for self)  [misc]
24
scipy/special/_ellip_harm.py:98: error: Too many arguments for "vectorize"  [call-arg]
25
scipy/special/_ellip_harm.py:98: error: Unexpected keyword argument "otypes" for "vectorize"  [call-arg]
26
/home/runner/.local/lib/python3.9/site-packages/mypy/typeshed/stdlib/3/builtins.pyi:93: note: "vectorize" defined here
27
scipy/special/_ellip_harm.py:161: error: Too many arguments for "vectorize"  [call-arg]
28
scipy/special/_ellip_harm.py:161: error: Unexpected keyword argument "otypes" for "vectorize"  [call-arg]
29
/home/runner/.local/lib/python3.9/site-packages/mypy/typeshed/stdlib/3/builtins.pyi:93: note: "vectorize" defined here
30
scipy/special/_ellip_harm.py:161: error: Incompatible types in assignment (expression has type "vectorize", variable has type "Callable[[Any, Any, Any, Any], Any]")  [assignment]
31
scipy/io/tests/test_idl.py:36: error: Too many arguments for "vectorize"  [call-arg]
32
scipy/fft/_pocketfft/tests/test_real_transforms.py:256: error: Invalid index type "Tuple[partial[Any], Type[int], int]" for "Dict[Tuple[partial[Any], Type[floating[Any]], int], int]"; expected type "Tuple[partial[Any], Type[floating[Any]], int]"  [index]
33
scipy/integrate/_ivp/common.py:7: error: Too many arguments for "finfo"  [call-arg]
34
scipy/sparse/linalg/isolve/lsqr.py:58: error: Too many arguments for "finfo"  [call-arg]
35
scipy/stats/_tukeylambda_stats.py:39: error: Too many arguments for "poly1d"  [call-arg]
36
scipy/stats/_tukeylambda_stats.py:40: error: Too many arguments for "poly1d"  [call-arg]
37
scipy/stats/_tukeylambda_stats.py:143: error: Too many arguments for "poly1d"  [call-arg]
38
scipy/stats/_tukeylambda_stats.py:144: error: Too many arguments for "poly1d"  [call-arg]
39
scipy/sparse/linalg/isolve/minres.py:370: error: No overload variant of "arange" matches argument types "int", "int", "Type[float]"  [call-overload]
40
scipy/sparse/linalg/isolve/minres.py:370: note: Possible overload variants:
41
scipy/sparse/linalg/isolve/minres.py:370: note:     def arange(stop: Any, dtype: Any = ..., *, like: Any = ...) -> Any
42
scipy/sparse/linalg/isolve/minres.py:370: note:     def arange(start: Any, stop: Any, step: Any, dtype: Any = ..., *, like: Any = ...) -> Any
43
scipy/sparse/linalg/isolve/minres.py:371: error: No overload variant of "arange" matches argument types "int", "int", "Type[float]"  [call-overload]
44
scipy/sparse/linalg/isolve/minres.py:371: note: Possible overload variants:
45
scipy/sparse/linalg/isolve/minres.py:371: note:     def arange(stop: Any, dtype: Any = ..., *, like: Any = ...) -> Any
46
scipy/sparse/linalg/isolve/minres.py:371: note:     def arange(start: Any, stop: Any, step: Any, dtype: Any = ..., *, like: Any = ...) -> Any
47
scipy/linalg/matfuncs.py:23: error: Too many arguments for "finfo"  [call-arg]
48
scipy/linalg/matfuncs.py:24: error: Too many arguments for "finfo"  [call-arg]
49
scipy/linalg/decomp_schur.py:178: error: Too many arguments for "finfo"  [call-arg]
50
scipy/linalg/decomp_schur.py:179: error: Too many arguments for "finfo"  [call-arg]
51
scipy/linalg/decomp.py:30: error: Invalid index type "str" for "Dict[generic, Callable[..., ndarray[Any, dtype[Any]]]]"; expected type "generic"  [index]
52
scipy/optimize/_lsq/common.py:12: error: Too many arguments for "finfo"  [call-arg]
53
scipy/linalg/tests/test_basic.py:26: error: Unsupported operand types for + ("List[Type[object]]" and "List[Type[object]]")  [operator]
54
scipy/optimize/slsqp.py:32: error: Too many arguments for "finfo"  [call-arg]
55
scipy/optimize/_trustregion_exact.py:211: error: Too many arguments for "finfo"  [call-arg]
56
scipy/optimize/_differentialevolution.py:19: error: Too many arguments for "finfo"  [call-arg]
57
scipy/optimize/tests/test_zeros.py:22: error: Too many arguments for "finfo"  [call-arg]
58
scipy/optimize/tests/test_zeros.py:24: error: Too many arguments for "finfo"  [call-arg]
59
scipy/stats/_qmc.py:454: error: Slice index must be an integer or None  [misc]
60
scipy/stats/_qmc.py:788: error: Argument 1 to "reshape" of "_ArrayOrScalarCommon" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
61
scipy/stats/_qmc.py:788: error: Argument 2 to "reshape" of "_ArrayOrScalarCommon" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
62
scipy/stats/_qmc.py:884: error: Incompatible types in assignment (expression has type "ndarray[Any, dtype[floating[_64Bit]]]", variable has type "float")  [assignment]
63
scipy/stats/_qmc.py:887: error: Argument 1 to "range" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
64
scipy/stats/_qmc.py:1056: error: Item "RandomState" of "Union[Generator, RandomState]" has no attribute "integers"  [union-attr]
65
scipy/stats/_qmc.py:1058: error: Item "Generator" of "Union[Generator, RandomState]" has no attribute "randint"  [union-attr]
66
scipy/stats/_qmc.py:1098: error: Slice index must be an integer or None  [misc]
67
scipy/stats/_qmc.py:1292: error: Incompatible return value type (got "Union[Any, bool_, ndarray[Any, dtype[bool_]]]", expected "ndarray[Any, Any]")  [return-value]
68
scipy/stats/_qmc.py:1325: error: Argument 1 to "reshape" of "_ArrayOrScalarCommon" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]
69
scipy/stats/_qmc.py:1327: error: Slice index must be an integer or None  [misc]
70
scipy/special/tests/test_orthogonal.py:114: error: Too many arguments for "poly1d"  [call-arg]
71
scipy/special/tests/test_orthogonal.py:140: error: Too many arguments for "poly1d"  [call-arg]
72
scipy/special/tests/test_orthogonal.py:165: error: Too many arguments for "poly1d"  [call-arg]
73
scipy/special/tests/test_orthogonal.py:190: error: Too many arguments for "poly1d"  [call-arg]
74
scipy/special/tests/test_orthogonal.py:215: error: Too many arguments for "poly1d"  [call-arg]
75
scipy/special/tests/test_orthogonal.py:266: error: Too many arguments for "poly1d"  [call-arg]
76
scipy/special/tests/test_orthogonal.py:266: error: "poly1d" not callable  [operator]
77
scipy/special/tests/test_orthogonal.py:274: error: Too many arguments for "poly1d"  [call-arg]
78
scipy/special/tests/test_orthogonal.py:275: error: Too many arguments for "poly1d"  [call-arg]
79
scipy/stats/tests/test_stats.py:63: error: Too many arguments for "finfo"  [call-arg]
```

BvB93 commented:
> Some Updates
> New errors are appeared out of nowhere. I wonder if it is due to new commit on NumPy's main branch...

Ugh, looking at the log there are a number of issues that have to be fixed in numpy.
Besides them, there are also few others that stand out:

> scipy/linalg/tests/test_basic.py:26: error: Unsupported operand types for + ("List[Type[object]]" and "List[Type[object]]")  [operator]

Honestly? I'm pretty sure this is a mypy bug, as addition of two lists of the same type should most definitely be considered legal.

> scipy/stats/_qmc.py:887: error: Argument 1 to "range" has incompatible type "Union[int, integer[Any]]"; expected "int"  [arg-type]

This actually appears to be an issue with the `range` definition in typeshed, which by the looks of it _should_ be able to take any object implementing the `__index__` protocol; not just builtin integers.

> scipy/stats/_qmc.py:1056: error: Item "RandomState" of "Union[Generator, RandomState]" has no attribute "integers"  [union-attr]

``` python
try:
    rg_integers = self.rng.integers
except AttributeError:
    rg_integers = self.rng.randint
```

Mypy currently does not support type narrowing via `try` / `except` blocks, hence the error.
Adding a `# type: ignore` is probably the best easiest solution here.

ilayn commented:
> Honestly? I'm pretty sure this is a mypy bug, as addition of two lists of the same type should most definitely be considered legal.

Unfortunately it is not a bug but by design, see links here https://github.com/scipy/scipy/pull/13876#issuecomment-821835941

BvB93 commented:
> Unfortunately it is not a bug but by design, see links here #13876 (comment)

The linked examples talk about `Sequence` (which does not define `__add__`), while the specific error talks about an invalid `+` operation between two lists (which _do_ define `__add__`). Oddly enough, I'm unable to reproduce the error with a minimal mockup example.

``` python
from typing import List, Type, TYPE_CHECKING

a: List[Type[object]]
b: List[Type[object]]

if TYPE_CHECKING:
    # Revealed type is 'builtins.list[Type[builtins.object]]'
    reveal_type(a + b)
```

ilayn commented:
Same with string examples in those links, their content is not of the same type. Hence cannot take them as just lists. That's why the error doesn't just say I don't know how to add Lists, but shows them as separate operands

BvB93 commented:
> scipy/stats/_qmc.py:887: error: Argument 1 to "range" has incompatible type "Union[int, integer[Any]]"; expected "int" [arg-type]

An update on the `range`-related error: It seems that problem has already been resolved earlier this year (https://github.com/python/typeshed/pull/4929). This error should thus disappear in the next mypy release.

V0lantis commented:
> > scipy/stats/_qmc.py:887: error: Argument 1 to "range" has incompatible type "Union[int, integer[Any]]"; expected "int" [arg-type]
> 
> An update on the `range`-related error: It seems that problem has already been resolved earlier this year ([python/typeshed#4929](https://github.com/python/typeshed/pull/4929)). This error should thus disappear in the next mypy release.

This is why I am proposing the changes in #13876, to take the latest Nightly Numpy release instead of the last release.

V0lantis commented:
Once all tests will have passed, it will be ready to be merged for me

V0lantis commented:
@BvB93 could you review please, for me it is green and ready to be merged :-)

BvB93 commented on scipy/stats/_qmc.py:
```diff
@@ -567,7 +619,12 @@ class QMCEngine(ABC):
+            *,
+            seed: SeedType = None
+    ) -> None:
         if not np.issubdtype(type(d), np.integer):
```
Unrelated, but FYI: `np.timdelta64` will also pass through this check.

BvB93 reviewed (commented):
Looks good overall, though annotations are still missing in a number of important places:

* `check_random_state` function, or all attributes taking its output will need an annotation
* `n_primes` function, or all attributes taking its output will need an annotation
* `Sobol.fast_forward` method
* `Sobol.MAXDIM` attribute (class variable, so needs a `typing.ClassVar`)
* `Sobol.MAXBIT` attribute (class variable, so needs `typing.ClassVar`)

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -148,7 +186,12 @@ def scale(sample, l_bounds, u_bounds, *, reverse=False):
+        *,
+        iterative: bool = False,
+        method: str = "CD",
+        workers: int = 1) -> float:
```
Shouldn't we use `IntegerNumber` and `DecimalNumber` everywhere?

BvB93 replied:
Both `range` and `list` can currently take any object implementing the `__index__` protocol, 
so using `n: np.integer = ...` should be fine.

Note that the most recent mypy release still has the `range` input types defined `builtins.int`, a more
recent typeshed update has loosened their types to aforementioned protocol:
https://github.com/python/typeshed/blob/740f67245033a4e647706869a01f67d8c079fd7f/stdlib/builtins.pyi#L904-L911

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -22,10 +34,30 @@
+# Since Generator was introduced in numpy 1.17, the following condition is needed for
+# backward compatibility
+if TYPE_CHECKING:
+    SeedType = Optional[Union[IntNumber, np.random.Generator,
```
I feel like these definitions should be in `utils` as we are likely going to need them everywhere.

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -1187,7 +1258,7 @@ def __init__(self, mean, cov=None, *, cov_root=None, inv_transform=True,
         self._corr_matrix = cov_root
 
-    def random(self, n=1):
+    def random(self, n: IntNumber = 1) -> np.ndarray:
```
We are inconsistent here as well with `not.ArrayLike` and `no.ndarray`.

V0lantis commented:
> Looks good overall, though annotations are still missing in a number of important places:
> 
> * `check_random_state` function, or all attributes taking its output will need an annotation

Are you sure of that ? I thought it was the purpose of `@overload` decorator. Line 56 : 

```python
@overload
def check_random_state(seed: Optional[IntNumber] = ...) -> np.random.Generator:
    ...

@overload
def check_random_state(seed: GeneratorType) -> GeneratorType:
    ...


# Based on scipy._lib._util.check_random_state
def check_random_state(seed=None):
```

> * `n_primes` function, or all attributes taking its output will need an annotation

Thanks, updated in : 4ca169301440b796104e70c5de4b4ace0a00ac11

> * `Sobol.fast_forward` method

Updated : `def fast_forward(self, n: IntNumber) -> Sobol:`

> * `Sobol.MAXDIM` attribute (class variable, so needs `typing.ClassVar`)
> * `Sobol.MAXBIT` attribute (class variable, so needs `typing.ClassVar`)

Updated as well in 4ca169301440b796104e70c5de4b4ace0a00ac11

Thank for your review ! :-) @BvB93 

---

BvB93 commented on scipy/stats/_qmc.py:
```diff
@@ -379,7 +427,7 @@ def primes_from_2_to(n):
 
 
-def n_primes(n):
+def n_primes(n: int):
```
```suggestion
def n_primes(n: int) -> List[int]:
```
The return type is really the important part here.

BvB93 reviewed (commented):
> Are you sure of that ? I thought it was the purpose of `@overload` decorator. Line 56 :

You are absolutely correct here, I somehow missed it when glancing over the code...

BvB93 commented on scipy/stats/_qmc.py:
```diff
@@ -22,10 +36,30 @@
            'MultinomialQMC', 'MultivariateNormalQMC']
 
+IntNumber = Union[int, np.integer]
+DecimalNumber = Union[float, np.floating]
```
```suggestion
DecimalNumber = Union[float, np.floating, np.integer]
```
Note that `builtins.int` is considered a subtype of `builtins.float` (by static type checkers, that is), 
but the same does not hold true for `np.integer` and `np.floating`.
I'd suggest making this change, unless you very explicitly want to exclude `np.integer`.

tupui commented on scipy/stats/utils.py:
```diff
@@ -0,0 +1,19 @@
+from typing import (
```
What I meant was to put these in `scipy._util` so that these can be used in all modules.

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -290,7 +328,10 @@ def discrepancy(sample, *, iterative=False, method="CD", workers=1):
+def update_discrepancy(
+        x_new: npt.ArrayLike,
+        sample: npt.ArrayLike,
+        initial_disc: float) -> float:
```
This function is public so needs this as well
```suggestion
        initial_disc: DecimalNumber) -> float:
```

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -347,7 +388,7 @@ def update_discrepancy(x_new, sample, initial_disc):
 
 
-def primes_from_2_to(n):
+def primes_from_2_to(n: IntNumber) -> np.ndarray:
```
This function is private and only called by `n_primes` using `int`.
```suggestion
def primes_from_2_to(n: int) -> np.ndarray:
```

tupui commented on scipy/stats/_qmc.py:
```diff
@@ -379,7 +420,7 @@ def primes_from_2_to(n):
 
 
-def n_primes(n):
+def n_primes(n: int) -> List[int]:
```
This one is called by `Halton`, hence can have something else than plain `int`
```suggestion
def n_primes(n: IntNumber) -> List[int]:
```

V0lantis reviewed (commented):
Thank you again @tupui for your review :-). We are getting close (finally) 🙃

tupui reviewed (approved):
LGTM, thanks @V0lantis! This will help further typing (the discussion and also what you added in `_util`.

Side note: before merging though we should have a cleaner history. I would either rebase and squash some commits together, and change the messages to comply with our naming conventions (I guess for type hints this should be `MAINT: ...`). Or we squash merge and clean the single commit message (co-authors are preserved, and we remove unnecessary things).

BvB93 commented on scipy/stats/_qmc.py:
```diff
@@ -1,13 +1,26 @@
 
+if TYPE_CHECKING:
+    import numpy.typing as npt
+    from typing import Literal
```
```suggestion
    from typing_extensions import Literal
```

Note that the `Literal` import will currently fail for python 3.7 (as it does not exist in python 3.7!).

You'll need either a version check here _or_ an unconditional import from `typing_extensions`, 
the latter which exposes either a backport or a re-export of `Literal`, depending on the python version.

BvB93 replied:
Fyi the import will have to remain within the `if TYPE_CHECKING` block as `typing_extensions` is not a runtime dependency.

V0lantis replied:
Oh sorry, I misunderstood : 

>  *or* an unconditional import from `typing_extensions`

I though unconditional meant outside the `if TYPE_CHECKING`.
Thank you very much @BvB93

V0lantis replied:
It seems that Scipy doesn't have `typing_extensions`. 
The error can be seen [here](https://github.com/scipy/scipy/runs/2595468155?check_suite_focus=true#step:8:651) : 

```
scipy/stats/__init__.py:444: in <module>
    from . import qmc
scipy/stats/qmc.py:234: in <module>
    from ._qmc import *
scipy/stats/_qmc.py:16: in <module>
    from typing_extensions import Literal
E   ModuleNotFoundError: No module named 'typing_extensions'
```
I found a similar issue [here](https://github.com/facebookresearch/cc_net/issues/3). Should I add `typing_extensions` in scipy requirements or something? @rgommers ?

BvB93 replied:
The `ModuleNotFoundError` is from the CI run prior to https://github.com/scipy/scipy/pull/13833/commits/136845f59dd725e53d482e106b9939fcebbecb4b. 
The error should not re-appear again, as the module is now no longer imported during runtime.

tupui commented:
@BvB93 is this ok for you with the latest changes? (CI failure is unrelated and now fixed on master)

BvB93 reviewed (approved):
Yup, looks good to me.

tupui commented:
Thanks everyone for the reviews! 

@V0lantis let me know what you prefer for the git cleanup before I integrate your PR.
Also just seeing this now, could you merge the two `if TYPE_CHECKING` together? (non blocking but would be better).

V0lantis commented:
> Thanks everyone for the reviews!
> 
> @V0lantis let me know what you prefer for the git cleanup before I integrate your PR.
> Also just seeing this now, could you merge the two `if TYPE_CHECKING` together? (non blocking but would be better).

I have squashed commits into one, but I am not seeing `if TYPE_CHECKING` repeats...

BvB93 commented on scipy/stats/_qmc.py:
```diff
@@ -1,13 +1,27 @@
 
 import numpy as np
 
+if TYPE_CHECKING:
+    import numpy.typing as npt
+    from typing_extensions import Literal
```
Over here.

BvB93 commented on scipy/stats/_qmc.py:
```diff
@@ -22,10 +36,24 @@
            'QMCEngine', 'Sobol', 'Halton', 'LatinHypercube',
            'MultinomialQMC', 'MultivariateNormalQMC']
 
+if TYPE_CHECKING:
+    from scipy._lib._util import (
+        DecimalNumber, GeneratorType, IntNumber, SeedType
+    )
```
And over here.

tupui commented:
> > Thanks everyone for the reviews!
> 
> > 
> 
> > @V0lantis let me know what you prefer for the git cleanup before I integrate your PR.
> 
> > Also just seeing this now, could you merge the two `if TYPE_CHECKING` together? (non blocking but would be better).
> 
> 
> 
> I have squashed commits into one, but I am not seeing `if TYPE_CHECKING` repeats...

Thanks for the cleanup! I will nitpick a bit now sorry, but could you edit the message to add @BvB93 as a co-author at least. It's perfect to squash but we should try to keep this co-authoring as mush as we can 😃

V0lantis commented:
Thank you both for your patience and your thoughtful comments 🙏

tupui reviewed (approved):
Thank you @V0lantis and @BvB93 for all the work!

If anyone has time, it would be really helpful to start a page in the wiki. I can do a draft if you guys don't have time. Let me know 😃

Merging once it's green.
